domains/emissions/Emissions.API/API/Services/SourcesCalculator.cs

```diff
@@ -0,0 +1,152 @@
+                        DateTo = measurement.DateTo,
+                        ProductionType = totalShare.ProductionType
+                    };
+                    shares.Add(totalShare.ProductionType, share);
+                }
+
+                share.Value += (float)totalShare.ShareTotal * measurement.Quantity;
+                if (measurement.DateFrom < share.DateFrom)
+                {
+                    share.DateFrom = measurement.DateFrom;
+                }
+
+                if (measurement.DateTo > share.DateTo)
+                {
+                    share.DateTo = measurement.DateTo;
+                }
+            }
+        }
+
+        EnergySourceResponse CalculateSourceEmissionPercentage(
+            IEnumerable<TimeSeries> timeSeries,
+            Aggregation aggregation,
+            Dictionary<string, Dictionary<string, ConsumptionShare>> consumptionResults)
+        {
+            var result = new EnergySourceResponse(new List<EnergySourceDeclaration>());
+            var groupedMeasurements = GetGroupedMeasurements(aggregation, timeSeries);
+
+            foreach (var consumptionResult in consumptionResults)
+            {
+                var matchingConsumptionSum = groupedMeasurements[consumptionResult.Key].Sum(_ => _.Quantity);
+
+                var sources = consumptionResult.Value.ToDictionary(a =>
+                    a.Key, b => (float)Math.Round((b.Value.Value / matchingConsumptionSum / 100), 5));
+
+                result.EnergySources.Add(new EnergySourceDeclaration
+                (
+                    consumptionResult.Value.Min(a => a.Value.DateFrom),
+                    consumptionResult.Value.Max(a => a.Value.DateTo),
+                    CalculateRenewable(sources),
+                    sources
+                ));
+            }
+
+            return result;
+        }
+
+        float CalculateRenewable(IDictionary<string, float> groupValues) =>
+             groupValues.Where(a => renewableSources.Contains(a.Key)).Sum(a => a.Value * (a.Key == "Waste" ? wasteRenewableShare : 1));
+
+        ILookup<string, Measurement> GetGroupedMeasurements(Aggregation aggregation, IEnumerable<TimeSeries> timeSeries)
+        {
+            if (aggregation == Aggregation.Total)
+                return timeSeries.SelectMany(a => a.Measurements).ToLookup(x => "total");
+
+            return timeSeries.SelectMany(y => y.Measurements)
+                .ToLookup(x => GetDateAsString(x.DateFrom.ToUtcDateTime(), aggregation));
+        }
+
+        ILookup<string, Record> GetGroupedDeclarations(Aggregation aggregation, List<Record> declaration)
+        {
+            if (aggregation == Aggregation.Total)
+                return declaration.ToLookup(a => "total" + a.PriceArea);
+
+            return declaration.ToLookup(a => GetDateAsString(a.HourUTC, aggregation) + a.PriceArea);
+        }
+
+        string GetDateAsString(DateTime date, Aggregation aggregation)
+        {
+            switch (aggregation)

```

Comment:
Consider making this switch less verbose, possibly with a switch expression:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression